src/transformers/models/audio_spectogram_transformer/modeling_audio_spectogram_transformer.py

```diff
@@ -0,0 +1,631 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch Audio Spectogram Transformer model."""
+
+import math
+from typing import Dict, List, Optional, Set, Tuple, Union
+
+import torch
+import torch.utils.checkpoint
+from torch import nn
+from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
+
+from ...activations import ACT2FN
+from ...modeling_outputs import BaseModelOutput, BaseModelOutputWithPooling, SequenceClassifierOutput
+from ...modeling_utils import PreTrainedModel
+from ...pytorch_utils import find_pruneable_heads_and_indices, prune_linear_layer
+from ...utils import add_code_sample_docstrings, add_start_docstrings, add_start_docstrings_to_model_forward, logging
+from .configuration_audio_spectogram_transformer import AudioSpectogramTransformerConfig
+
+
+logger = logging.get_logger(__name__)
+
+# General docstring
+_CONFIG_FOR_DOC = "AudioSpectogramTransformerConfig"
+_FEAT_EXTRACTOR_FOR_DOC = "AudioSpectogramTransformerFeatureExtractor"
+
+# Base docstring
+# TODO update to appropriate organization
+_CHECKPOINT_FOR_DOC = "nielsr/audio-spectogram-transformer-finetuned-audioset-10-10-0.4593"
+_EXPECTED_OUTPUT_SHAPE = [1, 1214, 768]
+
+# Audio classification docstring
+# TODO update to appropriate organization
+_SEQ_CLASS_CHECKPOINT = "nielsr/audio-spectogram-transformer-finetuned-audioset-10-10-0.4593"
+_SEQ_CLASS_EXPECTED_OUTPUT = "'Speech'"
+_SEQ_CLASS_EXPECTED_LOSS = 0.17
+
+
+AUDIO_SPECTOGRAM_TRANSFORMER_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    # TODO update to appropriate organization
+    "nielsr/audio-spectogram-transformer-finetuned-audioset-10-10-0.4593",
+    # See all Audio Spectogram Transformer models at https://huggingface.co/models?filter=audio-spectogram-transformer
+]
+
+
+class AudioSpectogramTransformerEmbeddings(nn.Module):
+    """
+    Construct the CLS token, position and patch embeddings.
+    """
+
+    def __init__(self, config: AudioSpectogramTransformerConfig) -> None:
+        super().__init__()
+
+        self.cls_token = nn.Parameter(torch.zeros(1, 1, config.hidden_size))
+        self.distillation_token = nn.Parameter(torch.zeros(1, 1, config.hidden_size))
+        self.patch_embeddings = AudioSpectogramTransformerPatchEmbeddings(config)
+
+        frequency_dimension, time_dimension = self.get_shape(config)
+        num_patches = frequency_dimension * time_dimension
+        self.position_embeddings = nn.Parameter(torch.zeros(1, num_patches + 2, config.hidden_size))
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+        self.config = config
+
+    def get_shape(self, config):
+        frequency_stride = config.frequency_stride
+        time_stride = config.time_stride
+        frequency_dimension = config.frequency_dimension
+        time_dimension = config.time_dimension
+        test_input = torch.randn(1, 1, frequency_dimension, time_dimension)
+        test_projection = nn.Conv2d(
```

Comment:
Is it strictly necessary to perform a 'dummy' convolution to get the downsampled feature dimensions? We can mathematically compute the number of downsampled dims from a conv layer given the kernel size and stride as:
$$W_{out} = \frac{W_{in}−F+2P}{S+1}$$
where $W_{in}$ is the number of input features, $W_{out}$ is the number of output features, $F$ the kernel size, $P$ the amount of zero-padding, and $S$ the stride (https://cs231n.github.io/convolutional-networks/#conv).

It seems a bit wasteful having to do a dummy fprop to compute $W_{out}$ when we can do it analytically!

Reply:
Thanks, indeed! Used this formula now.